src/plugins/rule/schema.js

```diff
@@ -0,0 +1,143 @@
+// Copyright 2019 OpenStax Poland
+// Licensed under the MIT license. See LICENSE file in the project root for
+// full license text.
+
+import { Block, Text } from 'slate'
+
+const TYPES = ['rule', 'theorem', 'lemma', 'corollary', 'law', 'proposition']
+
+function normalizeRule(editor, error) {
+    const { code, child, key, node, index } = error
+
+    switch (code) {
+    // A child of different type was expected.
+    case 'child_type_invalid': {
+        // A text child could only have been added to an exercise by Slate in
+        // event that all other children were removed. In such case we just want
+        // to drop the exercise.
+        if (Text.isText(child)) {
+            editor.removeNodeByKey(node.key)
+            return
+        }
+        // A non-empty rule without a statement should get an empty statement.
+        if (index === 0) {
+            const statement = Block.create({
+                type: 'rule_statement',
+                nodes: [
+                    // An empty leaf will be added by slate in a subsequent
+                    // normalization.
+                    Block.create({ type: 'paragraph' }),
+                ],
+            })
+            editor.insertNodeByKey(node.key, 0, statement)
+            return
+        }
+        // Otherwise (invalid child after the statement but not at the end of
+        // rule) we fold the invalid child into previous node.
+        const previous = node.nodes.get(index - 1)
+        editor.moveNodeByKey(child.key, previous.key, previous.nodes.size)
+        break
+    }
+
+    case 'child_min_invalid':
+        // A non-empty rule without a statement should get an empty statement.
+        if (index === 0) {
```

Comment:
`statement` is only child which can cause this error so there is no need to check `index`
I would suggest to first check if `rule_statement` is existing, but in wrong place and then if yes -> move it to correct position and if not then create new `statement`

Reply:
Ex. case like this:
```
export const input = <value>
    <document>
        <rule type="rule">
            <proof>
                <p>Missing proof</p>
            </proof>
            <statement>
                <p><text/></p>
            </statement>
        </rule>
    </document>
</value>
```
Will also trigger `child_min_invalid` and we shouldn't add new statement but just move this one to correct position

Reply:
I still prefer to make sure that the error is relevant. As for statements out of order, I've added a separate check and test for that case.